synapse/util/ratelimitutils.py

```diff
@@ -74,6 +74,27 @@ def new_limiter() -> "_PerHostRatelimiter":
             str, "_PerHostRatelimiter"
         ] = collections.defaultdict(new_limiter)
 
+        # We track the number of affected hosts per time-period so we can
+        # differentiate one really noisy homeserver from a general
+        # ratelimit tuning problem across the federation.
+        LaterGauge(
+            "synapse_rate_limit_sleep_affected_hosts",
+            "Number of hosts that had requests put to sleep",
+            [],
+            lambda: sum(
+                ratelimiter.should_sleep() for ratelimiter in self.ratelimiters.values()
+            ),
+        )
+        LaterGauge(
+            "synapse_rate_limit_reject_affected_hosts",
+            "Number of hosts that had requests rejected",
+            [],
+            lambda: sum(
+                ratelimiter.should_reject()
+                for ratelimiter in self.ratelimiters.values()
+            ),
```

Comment:
This seems to always evalulate to `0`.

Prometheus data: https://prometheus.matrix.org/graph?g0.expr=synapse_rate_limit_reject_affected_hosts&g0.tab=1&g0.stacked=0&g0.show_exemplars=0&g0.range_input=1h&g0.end_input=2022-08-22%2021%3A21%3A52&g0.moment_input=2022-08-22%2021%3A21%3A52

![](https://user-images.githubusercontent.com/558581/186021441-c12de1fd-52ef-4503-b05f-28a644e4ae1f.png)

Even though we see individual requests being rejected:

![](https://user-images.githubusercontent.com/558581/186021630-93202f39-c88d-4a34-8144-cf958e3643ce.png)

Reply:
As a note, when I log `should_reject()` in Complement, I see `should_reject queue_size=0 self.reject_limit=99999` (the high `reject_limit` is from [`docker/complement/conf/workers-shared-extra.yaml.j2#L36`](https://github.com/matrix-org/synapse/blob/3dd175b628bab5638165f20de9eade36a4e88147/docker/complement/conf/workers-shared-extra.yaml.j2#L36)) for all of the calls. Not sure I am exercising this enough in parallel to see it though.

Reply:
This is probably caused by https://github.com/matrix-org/synapse/issues/13641 which is being addressed in https://github.com/matrix-org/synapse/pull/13649

Reply:
I've tried to visualise this at https://grafana.matrix.org/d/000000012/synapse?orgId=1&viewPanel=225 but I still see 0 for all `rejected` metrics, even with #13649.

Reply:
These numbers look consistent with data I saw when investigating in the manhole (`sum(x.should_reject() for x in hs.get_federation_ratelimiter().ratelimiters.values())`). It might be that we need to wait some time for queues to grow before we start rejecting requests. I'd like to leave confirming/investigating this on your plate @MadLittleMods if that's ok?

Reply:
@DMRobertson Looks like https://github.com/matrix-org/synapse/pull/13649 hasn't fixed it if it has already been rolled out to `matrix.org` 😩 (still all 0)

Here are the rate limiter graph section I have:
https://grafana.matrix.org/d/dYoRgTgVz/messages-timing?orgId=1&from=1661854473240&to=1661876073240&viewPanel=220

I'll create an actual issue to track this more long-term now ⏩

Reply:
https://github.com/matrix-org/synapse/issues/13670 ⏩